Add preemption logic to GlobalState and typecheck

jvilk-stripe opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->

Add preemption logic to GlobalState and typecheck. The new logic is currently unused.

This change introduces two new classes:

* `TypecheckEpochManager`: Contains all of the logic related to slow path cancelation. Previously, all of this lived directly in `GlobalState`.
* `PreemptionTaskManager`: Contains all of the logic related to preemption. Passed into the `typecheck` method.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

This logic is needed to enable requests to preempt the slow path. The flow goes like this:

* The _typechecking thread_ is currently running a slow path.
* The _message processing thread_ attempts to preempt the slow path with a task (`trySchedulePreemptionTask`), which succeeds only if the slow path is still happening.
  * N.B.: `withEpochLock` protects against races between the slow path ending and a preemption task getting scheduled.
* The _typechecking thread_, within `pipeline::typecheck`, attempts to run a preemption task with each turn of the coordination loop (it wakes up with maximum latency ~20ms) using `tryRunScheduledPreemptionTask`.
  * If a task is registered, `tryRunScheduledPreemptionTask` grabs `typecheckMutex` as a _write lock_ (exclusive lock). Worker threads that are typechecking individual files conduct their work while holding a _read lock_ (shared lock) on `typecheckMutex` via `lockPreemption`. These threads periodically reacquire the lock, and will block if a thread is attempting to acquire a write lock. Thus, the preemption task only begins once all worker threads are blocked and are not performing any work.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

I added automated tests for the core GlobalState logic. There's no easy way to test `typecheck` itself. I intend to add end-to-end tests for this mechanism once I hook up LSP to the preemption logic.

jvilk-stripe commented on main/pipeline/pipeline.h:
```diff
@@ -31,8 +31,11 @@ std::vector<ast::ParsedFile> incrementalResolve(core::GlobalState &gs, std::vect
+// TODO(jvilk): Make this interface cleaner.
 ast::ParsedFilesOrCancelled typecheck(std::unique_ptr<core::GlobalState> &gs, std::vector<ast::ParsedFile> what,
-                                      const options::Options &opts, WorkerPool &workers);
+                                      const options::Options &opts, WorkerPool &workers, bool cancelable = false,
```
`cancelable` is now an explicit parameter because the fast path calls `typecheck` but is not cancelable.

jvilk-stripe commented:
cc @DarkDimius -- you might have strong opinions on this part of the code, as it impacts the core GlobalState abstraction.

dmitry-stripe commented on common/concurrency/WorkerPool.h:
```diff
@@ -11,6 +11,11 @@ class WorkerPool {
+    // [IDE] We want blocking operations to be more responsive to preemption requests.
+    inline static constexpr std::chrono::milliseconds PREEMPTION_BLOCK_INTERVAL() {
+        using namespace std::chrono_literals;
+        return 20ms;
```
are there any downsides of changing the other value from 250ms to 20ms?

jvilk-stripe replied:
Good question. I can time some runs on pay-server to see if it impacts throughput.

jvilk-stripe replied:
Nope! Perf seems unaffected running typecheck on pay-server.

DarkDimius commented on core/GlobalState.cc:
```diff
@@ -49,11 +50,11 @@ const int Symbols::MAX_PROC_ARITY;
 GlobalState::GlobalState(shared_ptr<ErrorQueue> errorQueue, shared_ptr<absl::Mutex> epochMutex,
                          shared_ptr<atomic<u4>> currentlyProcessingLSPEpoch, shared_ptr<atomic<u4>> lspEpochInvalidator,
-                         shared_ptr<atomic<u4>> lastCommittedLSPEpoch)
+                         shared_ptr<atomic<u4>> lastCommittedLSPEpoch, shared_ptr<shared_ptr<lsp::Task>> preemptTask)
```
I don't currently understand why would it be a constructor argument, could you give more context?

DarkDimius replied:
Also, do we expect Tasks to be immutable/threadsfe? The current code will share pointers between multiple global states(and thus threads)

DarkDimius replied:
also, should it be `atomic<shared_ptr<lsp::Task>>`?

jvilk-stripe replied:
> Also, do we expect Tasks to be immutable/threadsfe? The current code will share pointers between multiple global states(and thus threads)

Tasks themselves aren't threadsafe to run, but updating the `shared_ptr` atomically is threadsafe. I share the `shared_ptr` between GlobalStates so that we can update the task on `initialGS` in the message processing thread and it'll propagate to `finalGS` in the typechecking thread. It's similar to how initialGS and finalGS both share epoch counters.

> I don't currently understand why would it be a constructor argument, could you give more context?

I need to ensure that all GlobalStates from the same `initialGS` have the same `preemptTask` shared pointer.

> also, should it be atomic<shared_ptr<lsp::Task>>?

I recall having trouble using this formulation. Instead, I'm using the `shared_ptr` specializations of `atomic_load`/`atomic_store`:

https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

DarkDimius replied:
>. Tasks themselves aren't threadsafe to run

in the current api nothing a single task from being invoked from multiple threads concurrently.

DarkDimius commented on core/GlobalState.cc:
```diff
@@ -1492,7 +1493,9 @@ bool GlobalState::wasModified() const {
-    return lspEpochInvalidator->load() != currentlyProcessingLSPEpoch->load();
+    // Can't cancel typechecking if we have a preemption task to run -- we have to make it past resolver at least before
+    // we can cancel.
+    return lspEpochInvalidator->load() != currentlyProcessingLSPEpoch->load() && !atomic_load(preemptTask.get());
```
1. can we make a `nullptr` comparison explicit
2. I didn't understand how does checking for `preemptTask` being not null relate to comment of 'we have to make it past resolver at least before we can cancel'. Are there some backed-in assumptions here?

jvilk-stripe replied:
1. Sure!
2. I see the confusion here; I can update the comment to remove the resolver bit. "Can't cancel typechecking if we have a preemption task to run" captures the behavior completely. I wrote the resolver comment because we don't run preemption tasks until after resolver, and it would cause bugs if we canceled before resolver and _then_ ran the preeemption task.

DarkDimius commented on core/GlobalState.h:
```diff
@@ -264,6 +283,9 @@ class GlobalState final {
     const std::shared_ptr<std::atomic<u4>> lastCommittedLSPEpoch;
+    // Task to run when pre-empting typechecking. Outer shared_ptr is shared amongst all GlobalState instances that
+    // share a common lineage, whereas contents of inner shared_ptr is atomically replaced during preemption.
+    const std::shared_ptr<std::shared_ptr<lsp::Task>> preemptTask;
```
It's not currently clear for me why is the task ever storred into a field, could you elaborate?
Can it live only on stack?

jvilk-stripe replied:
I'm a bit confused by this comment. What's the alternative?

DarkDimius commented on core/GlobalState.h:
```diff
@@ -227,6 +240,12 @@ class GlobalState final {
+    // pre-empt.
+    // - Typechecking coordinator thread grabs as a writer lock when there's a preemption function, which halts all
+    // worker threads.
+    const std::unique_ptr<absl::Mutex> typecheckMutex;
```
This should probably be operated on via `Unfreeze*` classes rather than manually

DarkDimius commented on main/pipeline/pipeline.cc:
```diff
@@ -940,17 +945,28 @@ ast::ParsedFilesOrCancelled typecheck(unique_ptr<core::GlobalState> &gs, vector<
+                            if (preemptible) {
+                                // Acquire a reader lock here. Parks the thread if the typechecker thread is trying to
+                                // grab the lock to preempt typechecking.
+                                lock = make_unique<absl::ReaderMutexLock>(typecheckMutex.get());
```
The fact that user needs to remember to correctly grab locks bothers me.
Can we make an abstraction where api forces users to use it right?
https://github.com/sorbet/sorbet/pull/2407/files#r362692618

jvilk-stripe replied:
So worker thread Unfreeze's GlobalState before doing work which 'does the right thing' and grabs the reader lock?

jvilk-stripe replied:
I can look into how the existing Unfreeze classes work.

jvilk-stripe commented on core/GlobalState.cc:
```diff
@@ -1491,85 +1491,6 @@ bool GlobalState::wasModified() const {
     return wasModified_;
 }
 
-bool GlobalState::wasTypecheckingCanceled() const {
```
Moved to `TypecheckingEpochManager`.

DarkDimius replied:
Awesome cleanup

jvilk-stripe commented on core/GlobalState.h:
```diff
@@ -250,21 +239,6 @@ class GlobalState final {
     bool wasModified_ = false;
 
-    // In LSP mode: Used to linearize operations involving lastCommittedLSPEpoch.
-    const std::shared_ptr<absl::Mutex> epochMutex;
```
Moved to `TypecheckingEpochManager`

jvilk-stripe commented on core/lsp/TypecheckEpochManager.cc:
```diff
@@ -0,0 +1,116 @@
+using namespace std;
+namespace sorbet::core::lsp {
+
+void TypecheckEpochManager::assertConsistentThread(optional<thread::id> &expectedThreadId, string_view method,
```
We don't know all of the thread IDs at construction time, so this assertion primarily asserts that we call specific routines from a consistent thread ID.

jvilk-stripe commented on core/lsp/TypecheckEpochManager.cc:
```diff
@@ -0,0 +1,116 @@
+    if (preemptionManager.has_value()) {
+        // Now that we are no longer running a slow path, run a preemption task that might have snuck in while we were
+        // finishing up. No others can be scheduled.
+        (*preemptionManager)->tryRunScheduledPreemptionTask();
```
I'm not 100% happy that `TypecheckEpochManager` knows about `PreemptionTaskManager`, but I don't have a better solution at the moment.

jvilk-stripe commented on main/pipeline/pipeline.cc:
```diff
@@ -915,11 +917,19 @@ ast::ParsedFilesOrCancelled resolve(unique_ptr<core::GlobalState> &gs, vector<as
 ast::ParsedFilesOrCancelled typecheck(unique_ptr<core::GlobalState> &gs, vector<ast::ParsedFile> what,
-                                      const options::Options &opts, WorkerPool &workers) {
+                                      const options::Options &opts, WorkerPool &workers, bool cancelable,
+                                      optional<shared_ptr<core::lsp::PreemptionTaskManager>> preemptionManager) {
```
I could make this non-optional and default to a new preemption manager (so that preemption will never actually occur), but that would add the overhead of a reader lock acquisition to every file.

jvilk-stripe commented:
I believe I've addressed the feedback you've given me, @DarkDimius. I'm working on benchmarking Sorbet w/ the 20ms change.

Other than that, this is ready for review!

DarkDimius commented on core/lsp/PreemptionTaskManager.h:
```diff
@@ -0,0 +1,41 @@
+    // - Worker threads grab as a reader lock, and routinely gives up and re-acquire the lock to allow other requests to
+    // pre-empt.
+    // - Typechecking coordinator thread grabs as a writer lock when there's a preemption function, which halts all
+    // worker threads.
```
```suggestion
    // worker threads. Only typechecking coordinator can do this and this enforces that multiple threads can't get mutable access to global state.
```

jvilk-stripe commented:
The 20ms change has no measurable impact on performance. Typechecking pay-server (on my branch on my devbox, which has some typechecking failures) takes ~24s with and without the change.
